play-json/src/main/scala/com/lightbend/lagom/scaladsl/playjson/PlayJsonSerializer.scala

```diff
@@ -66,16 +67,17 @@ private[lagom] final class PlayJsonSerializer(val system: ExtendedActorSystem, r
   override def fromBinary(bytes: Array[Byte], manifest: String): AnyRef = {
     val startTime = if (isDebugEnabled) System.nanoTime else 0L
 
-    val i = manifest.lastIndexOf('#')
-    val fromVersion = if (i == -1) 1 else manifest.substring(i + 1).toInt
-    val manifestClassName = if (i == -1) manifest else manifest.substring(0, i)
+    val (fromVersion: Int, manifestClassName: String) = parseManifest(manifest)
 
     val migration = migrations.get(manifestClassName)
 
     val migratedManifest = migration match {
-      case Some(migration) if migration.currentVersion > fromVersion =>
+      case Some(migration) if (migration.currentVersion > fromVersion) =>
         migration.transformClassName(fromVersion, manifestClassName)
-      case None => manifestClassName
+      case Some(migration) if (migration.currentVersion < fromVersion) =>
+        throw new IllegalStateException(s"Migration version ${migration.currentVersion} is " +
+          s"behind version $fromVersion of deserialized type [$manifestClassName]")
```

Comment:
Thank you James for the subtle points on deployment!
But does it mean that for the typical multi-node deploy (one micro-service lives in many nodes), we should still be safe because of the consistent hashing of entities assigned to the node given the same node configuration?
By that I mean, if nodeA is deployed first with new migration version (upgraded), the events persisted belonging to this new version will only be read by nodeA? And when subsequent nodes got upgraded, some of the entities previously belong to nodeA might now go to the new nodes, and that is fine as they have the new migration version as well